marker_api/src/diagnostic.rs

```diff
@@ -1,36 +1,74 @@
 //! This module is responsible for the construction of diagnostic messages. The
 //! [`DiagnosticBuilder`] is the public stable interface, to construct messages.
 
+use std::fmt::Debug;
+
 use crate::{
-    ast::{ExprId, FieldId, ItemId, Span, StmtId, VariantId},
-    context::AstContext,
+    ast::{Identifiable, NodeId, Span},
+    context::{with_cx, AstContext},
     ffi::{FfiSlice, FfiStr},
     lint::Lint,
+    prelude::Spanned,
 };
 
 /// This builder creates the diagnostic object which will be emitted by the driver.
 /// The documentation will showcase the messages in rustc's console emission style,
 /// the actual display depends on the driver.
 pub struct DiagnosticBuilder<'ast> {
+    /// This field will be `Some` if the created diagnostic will be emitted, otherwise
+    /// it'll be `None`. See [`DiagnosticBuilder::decorate`] for more information, when the
+    /// lint might be suppressed.
+    inner: Option<DiagnosticBuilderInner<'ast>>,
+}
+
+struct DiagnosticBuilderInner<'ast> {
     lint: &'static Lint,
+    node: NodeId,
     msg: String,
-    node: EmissionNode,
     span: Span<'ast>,
     parts: Vec<DiagnosticPart<String, Span<'ast>>>,
 }
 
-#[allow(clippy::needless_pass_by_value)] // `&impl ToString` doesn't work
+#[allow(clippy::needless_pass_by_value)] // `&impl Into<String>` doesn't work
```

Comment:
I'm not sure what you meant by this justification comment originally for `&imp ToString` doesn't work. What exactly doesn't work with it?

..but `&impl Into<String>` is definitely not going to work, because `Into<String>` consumes `self`, and when you have an `&impl Into<String>` you can't consume it, so if you ever happen to stumble with this type, then this type is totally useless, because you won't be able to call the only `into(self)` method available for it.

Anyway, I checked that if you remove this `allow()` clippy still passes. I suppose it's because there is no longer any `impl ToString` used in this impl block.